src/simulator/halo_simulator.py

```diff
@@ -243,40 +243,54 @@ def simulated_venn_diagram_reach_by_spend(
-              spent with publisher i. Note that publishers with 0 spends will
-              not be included in the Venn diagram reach.
+            spends:  The hypothetical spend vector, equal in length to the
+              number of publishers.  spends[i] is the amount that is spent with
+              publisher i. Note that the publishers with 0 spends, inactive
+              publishers, will not be included in the Venn diagram regions.
             max_frequency:  The maximum frequency for which to report reach.
         Returns:
-            regions:  A dictionary in which each key are the binary
+            regions:  A dictionary in which each key is the binary
               representations of each primitive region of the Venn diagram, and
-              each value is a list of the k+ reaches in the corresponding region.
-              The k+ reach for a given region is given as a list r[] where r[k]
-              is the number of people who were reached AT LEAST k+1 times.
+              each value is a list of the k+ reaches in the corresponding
+              region.
+              Note that the binary representation of a key represents the
+              formation of publisher IDs in that primitive region. For example,
+              primitive_regions[key] with key = 5 = bin('101') is the region
+              which belongs to pub_id-0 and id-2.
+              The k+ reaches for a given region is given as a list r[] where
+              r[k] is the number of people who were reached AT LEAST k+1 times.
         """
         # Get user counts by spend for each active publisher
-        user_counts_by_pub_id = {}
-        for pub_id, spend in enumerate(spends):
-            if not spend:
-                continue
-            user_counts_by_pub_id[pub_id] = self._publishers[
-                pub_id
-            ]._publisher_data.user_counts_by_spend(spend)
+        user_counts_by_pub_id = {
+            pub_id: self._publishers[pub_id]._publisher_data.user_counts_by_spend(spend)
+            for pub_id, spend in enumerate(spends)
+            if spend
+        }
 
         if len(user_counts_by_pub_id) > MAX_ACTIVE_PUBLISHERS:
             raise ValueError(
                 f"There are {len(user_counts_by_pub_id)} publishers for the Venn "
                 f"diagram algorithm. The maximum limit is {MAX_ACTIVE_PUBLISHERS}."
             )
 
-        # Locate user's region represented by a number and sum the impressions.
+        # Generate the representations of all primitive regions from the
+        # powerset of the active publishers, excluding the empty set.
+        active_pubs = list(user_counts_by_pub_id.keys())
+        active_pub_powerset = chain.from_iterable(
+            combinations(active_pubs, r) for r in range(1, len(active_pubs) + 1)
+        )
+        regions_with_active_pubs = [
+            sum(1 << pub_id for pub_id in pub_ids) for pub_ids in active_pub_powerset
+        ]
```

Comment:
The codes here are very elegant, but not easy to read.  Perhaps you can give an example to show what `active_pubs`, `active_pub_powerset` and `regions_with_active_pubs` look like.

Reply:
Thanks for the suggestion. Added example in the comment.